receiver/apachesparkreceiver/scraper.go

```diff
@@ -0,0 +1,332 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/apachesparkreceiver/internal/metadata"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/apachesparkreceiver/internal/models"
+)
+
+var (
+	errClientNotInit         = errors.New("client not initialized")
+	errFailedAppIDCollection = errors.New("failed to retrieve app ids")
+	errNoMatchingWhitelistedApps     = errors.New("no apps matched whitelisted names")
+)
+
+type sparkScraper struct {
+	client   client
+	logger   *zap.Logger
+	config   *Config
+	settings component.TelemetrySettings
+	mb       *metadata.MetricsBuilder
+}
+
+func newSparkScraper(logger *zap.Logger, cfg *Config, settings receiver.CreateSettings) *sparkScraper {
+	return &sparkScraper{
+		logger:   logger,
+		config:   cfg,
+		settings: settings.TelemetrySettings,
+		mb:       metadata.NewMetricsBuilder(cfg.MetricsBuilderConfig, settings),
+	}
+}
+
+func (s *sparkScraper) start(_ context.Context, host component.Host) (err error) {
+	httpClient, err := newApacheSparkClient(s.config, host, s.settings)
+	if err != nil {
+		return fmt.Errorf("failed to start: %w", err)
+	}
+	s.client = httpClient
+	return nil
+}
+
+func (s *sparkScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	now := pcommon.NewTimestampFromTime(time.Now())
+	var scrapeErrors scrapererror.ScrapeErrors
+
+	if s.client == nil {
+		return pmetric.NewMetrics(), errClientNotInit
+	}
+
+	// call applications endpoint to get ids and names for all apps in the cluster
+	var apps *models.Applications
+	apps, err := s.client.GetApplications()
+	if err != nil {
+		return pmetric.NewMetrics(), errFailedAppIDCollection
+	}
+
+	// check apps against whitelisted app ids from config
+	var whitelistedApps []models.Application
+
+	// if no ids specified, whitelist all apps
+	switch {
+	case s.config.WhitelistedApplicationNames == nil:
+		whitelistedApps = *apps
+	case len(s.config.WhitelistedApplicationNames) == 0:
+		whitelistedApps = *apps
+		s.logger.Warn("Empty array of whitelisted application IDs specified - all applications will be monitored.")
+	default:
+		// some whitelisted ids specified, compare to ids from applications endpoint
+		for _, app := range *apps {
+			if slices.Contains(s.config.WhitelistedApplicationNames, app.Name) {
+				whitelistedApps = append(whitelistedApps, app)
+			}
+		}
+		if len(whitelistedApps) == 0 {
+			return pmetric.NewMetrics(), errNoMatchingWhitelistedApps
+		}
+	}
+
+	// get stats from the 'metrics' endpoint
+	clusterStats, err := s.client.GetClusterStats()
+	if err != nil {
+		scrapeErrors.AddPartial(1, err)
```

Comment:
the number associated with an addpartial should be the number of metrics lost, if it can be known.